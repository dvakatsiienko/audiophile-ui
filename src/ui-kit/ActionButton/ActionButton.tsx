/* Core */
import Image from 'next/future/image';
import NextLink from 'next/link';
import styled, { css } from 'styled-components';

/* Instruments */
import { center } from '@/ui-kit';
import type { Route } from '@/types';
import chevronRight from './img/chevron-right.svg';

export const ActionButton = (props: ActionButtonProps) => {
    const children = (
        <span>
            <span>{props.children}</span>
            {props.variant === 'inline' ? <Image alt = 'chevron' src = { chevronRight } /> : null}
        </span>
    );

    const handlePointerUp = () => {
        if (props.onPointerUp) {
            props.onPointerUp();
            return null;
        }
    };

    if (props.as === 'a') {
        return (
            <NextLink href = { props.href as unknown as URL }>
                <Link
                    $size = { props.size as Size }
                    $variant = { props.variant as Variant }
                    href = { props.href }
                >
                    {children}
                </Link>
            </NextLink>
        );
    }

    return (
        <Button
            $size = { props.size as Size }
            $variant = { props.variant as Variant }
            data-testid = { props[ 'data-testid' ] }
            type = { props.type }
            onPointerUp = { handlePointerUp }
        >
            {children}
        </Button>
    );
};
ActionButton.defaultProps = {
    onPointerUp:   () => null,
    children:      'Click',
    variant:       'primary',
    type:          'button',
    as:            'button',
    href:          '/',
    size:          'medium',
    'data-testid': null,
};

/* Styles */
const heights: Record<Size, string> = {
    small:  '20px',
    medium: '40px',
    large:  '48px',
};

const styles = css<SButtonProps>`
    display: flex;
    justify-content: center;
    align-items: center;
    min-width: 160px;
    height: ${p => heights[ p.$size ]};
    outline: none;
    border: none;
    transition: color 0.3s ease, background-color 0.3s ease;
    font-weight: 700;
    font-size: 13px;
    line-height: 17px;
    text-transform: uppercase;
    user-select: none;

    ${p => variants[ p.$variant as Variant ]}
`;

const Button = styled.button<SButtonProps>`
    ${styles}
`;
const Link = styled.a<SButtonProps>`
    ${styles}
`;

const variants = {
    primary: css`
        background-color: var(--color-6);
        color: var(--color-3);

        &:hover {
            background-color: var(--color-7);
        }
    `,
    secondary: css`
        background-color: var(--color-5);
        color: var(--color-1);
        border: 1px solid black;

        &:hover {
            background-color: var(--color-1);
            color: var(--color-5);
        }
    `,
    inline: css`
        background-color: transparent;
        color: var(--color-8);

        &:hover {
            color: var(--color-6);
        }

        & > span {
            ${center};
            flex-direction: row;

            display: flex;
            gap: 13px;
        }
    `,
};

/* Types */
type ActionButtonProps = React.PropsWithChildren<{
    onPointerUp?: () => void;
    variant?: Variant;
    // type?: string;
    type?: ButtonType;
    as?: As;
    href?: Route;
    size?: Size;
    ['data-testid']?: string;
}>;

type Variant = 'primary' | 'secondary' | 'inline';
type As = 'button' | 'a';
type Size = 'small' | 'medium' | 'large';
type ButtonType = React.ButtonHTMLAttributes<HTMLButtonElement>['type'];

interface SButtonProps {
    $variant: Variant;
    $size: Size;
}
