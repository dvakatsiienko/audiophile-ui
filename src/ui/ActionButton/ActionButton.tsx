/* Core */
import React from 'react';
import Image from 'next/future/image';
import styled, { css } from 'styled-components';

/* Instruments */
import chevronRight from './img/chevron-right.svg';
import type { Route } from '@/types';

export const ActionButton = (props: ActionButtonProps) => {
    const dynamicProps: { type?: string; href?: string } = {};

    if (props.as === 'button') dynamicProps.type = props.type;
    if (props.as === 'a') dynamicProps.href = props.href;

    const handlePointerUp = () => {
        if (props.onPointerUp) {
            props.onPointerUp();
            return null;
        }
    };

    return (
        <Button
            { ...dynamicProps }
            $size = { props.size as Size }
            $variant = { props.variant as Variant }
            as = { props.as }
            data-testid = { props[ 'data-testid' ] }
            onPointerUp = { handlePointerUp }
        >
            <span>{props.children}</span>
            {props.variant === 'inline' ? <Image alt = 'chevron' src = { chevronRight } /> : null}
        </Button>
    );
};
ActionButton.defaultProps = {
    onPointerUp:   () => null,
    children:      'Click',
    variant:       'primary',
    type:          'button',
    as:            'button',
    href:          '/',
    size:          'medium',
    'data-testid': null,
};

/* Styles */
const variants = {
    primary: css`
        background-color: var(--color-6);
        color: var(--color-3);

        &:hover {
            background-color: var(--color-7);
        }
    `,
    secondary: css`
        background-color: var(--color-5);
        color: var(--color-1);
        border: 1px solid black;

        &:hover {
            background-color: var(--color-1);
            color: var(--color-5);
        }
    `,
    inline: css`
        gap: 13px;
        background-color: transparent;
        color: var(--color-8);

        &:hover {
            color: var(--color-6);
        }
    `,
};

const heights: Record<Size, string> = {
    small:  '20px',
    medium: '40px',
    large:  '48px',
};

const Button = styled.button<SButtonProps>`
    display: flex;
    justify-content: center;
    align-items: center;
    width: 160px;
    height: ${p => heights[ p.$size ]};
    outline: none;
    transition: color 0.3s ease, background-color 0.3s ease;
    border: none;
    user-select: none;

    ${p => variants[ p.$variant as Variant ]}
`;

/* Types */
interface ActionButtonProps {
    onPointerUp?: () => void;
    children?: React.ReactNode;
    variant?: Variant;
    type?: 'button' | 'submit' | 'reset';
    as?: As;
    href?: Route;
    size?: Size;
    ['data-testid']?: string;
}

type Variant = 'primary' | 'secondary' | 'inline';
type As = 'button' | 'a';
type Size = 'small' | 'medium' | 'large';

interface SButtonProps {
    $variant: Variant;
    $size: Size;
}
