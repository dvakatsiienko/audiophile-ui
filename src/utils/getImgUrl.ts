/* Instruments */
import { API_URL } from './api-url';

export const getImgUrl = (url: string) => {
    const result = `${API_URL}${url}`;

    return result;
};
