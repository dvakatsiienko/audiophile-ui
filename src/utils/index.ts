export * from './api-url';
export * from './getImgUrl';
export * from './test-utils';
export * from './useClickOutside';
