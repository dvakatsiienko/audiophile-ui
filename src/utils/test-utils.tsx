/* Core */
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

/* Instruments */
import { theme as defaultTheme } from '@/ui-kit';

const styledRender = (ui: React.ReactElement, { theme = defaultTheme, ...options } = {}) => {
    const Wrapper = (props: React.PropsWithChildren) => {
        return <ThemeProvider theme = { theme }>{props.children}</ThemeProvider>;
    };

    return render(ui, { wrapper: Wrapper, ...options });
};

export { styledRender };
