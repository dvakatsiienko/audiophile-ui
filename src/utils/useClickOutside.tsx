/* Core */
import { useState, useEffect, type RefObject } from 'react';

export const useClickOutside = (ref: RefObject<HTMLElement>, cb: () => void, bool: boolean) => {
    const [ isMounted, setIsMounted ] = useState(false);

    useEffect(() => {
        setIsMounted(true);
    }, []);

    useEffect(() => {
        const handleClickOutside = (event: globalThis.MouseEvent) => {
            const { current } = ref;

            if (isMounted && current && !current.contains(event.target as unknown as Node | null)) {
                cb();
            }
        };

        document.addEventListener('mousedown', handleClickOutside);
        return () => document.removeEventListener('mousedown', handleClickOutside);
    }, [ ref, bool, isMounted ]);
};
