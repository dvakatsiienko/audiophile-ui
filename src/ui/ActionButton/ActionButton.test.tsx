/* Core */
import { screen } from '@testing-library/react';
import user from '@testing-library/user-event';

/* Components */
import { ActionButton } from './ActionButton';

/* Instruments */
import { styledRender } from '@/utils';

const handler = jest.fn();
jest.spyOn(ActionButton.defaultProps, 'onClick');
jest.spyOn(console, 'error');

beforeEach(jest.clearAllMocks);

describe.only('<ActionButton />:', () => {
    test('renders correctly primary variant', async () => {
        styledRender(<ActionButton />);
        const button = screen.getByRole('button');

        expect(button).toMatchSnapshot();
    });

    test('renders correctly secondary variant', async () => {
        styledRender(<ActionButton variant = 'secondary' />);
        const button = screen.getByRole('button');

        expect(button).toMatchSnapshot();
    });

    test('renders correctly inline variant', async () => {
        styledRender(<ActionButton variant = 'inline' />);
        const button = screen.getByRole('button');

        expect(button).toMatchSnapshot();
    });

    test('renders as <a />', async () => {
        const href = '/speakers';

        styledRender(<ActionButton as = 'a' href = { href } />);
        const button = screen.getByRole('link');

        expect(button).toHaveAttribute('href', href);
        expect(button.tagName).toBe('A');
    });

    test('handles click handler passed via props', async () => {
        styledRender(<ActionButton onClick = { handler } />);
        const button = screen.getByRole('button');

        await user.click(button);

        expect(handler).toBeCalledTimes(1);
    });

    test('provides default stub function via default props', async () => {
        styledRender(<ActionButton />);
        const button = screen.getByRole('button');

        await user.click(button);

        expect(ActionButton.defaultProps.onClick).toBeCalledTimes(1);
    });
});
