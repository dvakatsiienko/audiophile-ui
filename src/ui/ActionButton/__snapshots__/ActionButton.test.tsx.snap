// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionButton />: renders correctly inline variant 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 160px;
  height: 40px;
  outline: none;
  -webkit-transition: color 0.3s ease,background-color 0.3s ease;
  transition: color 0.3s ease,background-color 0.3s ease;
  border: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  gap: 13px;
  background-color: transparent;
  color: var(--color-8);
}

.c0:hover {
  color: var(--color-6);
}

<button
  class="c0"
  type="button"
>
  <span>
    Click
  </span>
  <img
    alt="chevron"
    data-nimg="future"
    decoding="async"
    height="100"
    loading="lazy"
    src="http://localhost/_next/static/media/mocked-path.svg"
    srcset="http://localhost/_next/static/media/mocked-path.svg 1x, http://localhost/_next/static/media/mocked-path.svg 2x"
    style="color: transparent;"
    width="100"
  />
</button>
`;

exports[`<ActionButton />: renders correctly primary variant 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 160px;
  height: 40px;
  outline: none;
  -webkit-transition: color 0.3s ease,background-color 0.3s ease;
  transition: color 0.3s ease,background-color 0.3s ease;
  border: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: var(--color-6);
  color: var(--color-3);
}

.c0:hover {
  background-color: var(--color-7);
}

<button
  class="c0"
  type="button"
>
  <span>
    Click
  </span>
</button>
`;

exports[`<ActionButton />: renders correctly secondary variant 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 160px;
  height: 40px;
  outline: none;
  -webkit-transition: color 0.3s ease,background-color 0.3s ease;
  transition: color 0.3s ease,background-color 0.3s ease;
  border: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: var(--color-5);
  color: var(--color-1);
  border: 1px solid black;
}

.c0:hover {
  background-color: var(--color-1);
  color: var(--color-5);
}

<button
  class="c0"
  type="button"
>
  <span>
    Click
  </span>
</button>
`;
